### ISSUE TITLE
Object Feature Extraction Based on Bounding Box Coordinates

### ISSUE THREAD

**Issue by gigumay** [ultralytics/ultralytics#24298]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, 
I am trying to extract object features from the YOLO11 feature maps. I do this by using the `embed` parameter of the prediction function to obtain feature maps. I then map the object  bounding box coordinates onto the feature map by dividing them by the stride. E.g.: `x1_fm = x1_im * 1/stride` where `x1_fm` is the feature map coordinate as computed from the coordinate in the input image `x1_im`.  I am doubtful whether that is too simplistic an approach? In the sense that from what I understand even if I turn off augmentations, the Letterbox class will still be invoked and modify the input. Now, I have pre-cropped my images to 640x640 meaning that no up- or downscaling should be necessary under the hood. Turning off all other augmentations, can I safely assume that my coordinates will be mapped onto the feature maps correctly?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @gigumay, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue moving quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a feature-extraction ❓ question, please provide as much information as possible to help us understand your setup, including:
-Your exact `predict()` call and `embed` usage
-The YOLO11 model variant you are using
-A small code snippet showing how you map bounding boxes to feature-map coordinates
-Example input image shape and any preprocessing steps
-Any relevant logs or printed tensor shapes

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can get it from Ultralytics by using the code here:

https://github.com/ultralytics/ultralytics/issues/22562#issuecomment-3476997981

**Comment by gigumay** [comment]
Thank you very much! The is very useful to extract features of predicted objects. I am however looking to extract features based on the bounding box coordinates of ground truth labels. Any thought on whether it could make sense to modify your proposed solution so that I can pass the coordinates as well?

**Comment by Y-T-G** [comment]
For that, the best approximation would be to use ROIAlign. But YOLO is not trained with ROIAlign, so the correspondence would  probably not be very accurate. The alternative is to get the vector of the anchor closest to the center of the bounding box. You would have feature from 3 FPN levels You could probably concatenate them to a single vector or choose to only use one of them.

**Comment by gigumay** [comment]
Thank you! I have experimented with RoIAlign and indeed it doesn't seem to work for what I have in mind. Are you suggesting that I consider only a single anchor per object? Wouldn't larger objects span multiple anchors?

**Comment by Y-T-G** [comment]
YOLO is designed to learn single anchor per object. The size is accounted for by multi scale FPN levels. For larger objects, you use the anchors from P5.

**Comment by gigumay** [comment]
I understand, thanks a lot for your help!

**Comment by glenn-jocher** [comment]
You're welcome — if you continue exploring GT-aligned feature extraction, please open a [GitHub Discussion](https://github.com/orgs/ultralytics/discussions) with a small reproducible example so the Ultralytics team can take a closer look.

**Comment by gigumay** [comment]
Sorry, I do have one more question: To find the anchor corresponding to the bounding box center, do you think `int((bbox_centre / stride)) - 0.5` will give me a good enough approximation?

**Comment by glenn-jocher** [comment]
Not quite — YOLO anchor points are centered at `grid + 0.5`, so use `idx = round(bbox_center / stride - 0.5)` to get the nearest cell index, or `anchor_center = (idx + 0.5) * stride` in image space; see `make_anchors(..., grid_cell_offset=0.5)` in the [TAL reference](https://docs.ultralytics.com/reference/utils/tal/).

**Comment by gigumay** [comment]
Thanks for your reply - In the end I ended up using the `fg_mask` and `target_gt_idx` returned from the `TaskAlignedAssignert`:

`class TALFeatureExtractor(nn.Module):
    """
    Extract object-level features from feature maps using TAL assignments.
    
    Flattens and concatenates feature maps across levels (matching make_anchors order),
    then pools assigned anchor features per GT object, optionally weighted by align_metric.
    """

    def __init__(self):
        super().__init__()

    def forward(
        self,
        embds: list,
        fg_mask: torch.Tensor,
        target_gt_idx: torch.Tensor,
    ) -> torch.Tensor:
        """
        Args:
            embds:          list of feature maps [(B, C, H_i, W_i), ...], same level 
                            order as strides passed to make_anchors.
            fg_mask:        (B, A) bool — foreground anchor mask from TAL.
            target_gt_idx:  (B, A) long — which GT each anchor is assigned to.

        Returns:
            obj_features:   (N_objects, C) — one pooled feature vector per GT object
                            across the batch. Returns empty tensor if no foreground anchors.
        """

        # flatten spatial dims and concat levels: (B, C, A_total)
        # flatten(2) is row-major (H, W) which matches make_anchors meshgrid(indexing='ij')
        all_feats = torch.cat([f.flatten(2) for f in embds], dim=2)

        bs = fg_mask.shape[0]
        obj_features = []

        for b in range(bs):
            pos_idx = fg_mask[b].nonzero(as_tuple=True)[0]  # (P,)
            if pos_idx.numel() == 0:
                continue

            gt_idx = target_gt_idx[b, pos_idx]    # (P,)
            feats_pos = all_feats[b, :, pos_idx]  # (C, P)

            for g in gt_idx.unique():
                sel_mask = gt_idx == g
                sel_feats = feats_pos[:, sel_mask]  # (C, K)

                obj_features.append(sel_feats.mean(dim=1))

        if not obj_features:
            return None

        return torch.stack(obj_features) `


Does this seem like an appropriate approach to you? As `fg_mask` should have already filtered anchors outside of GT bounding boxes, this should guarantee that I only get features from within GT labels, right?

**Comment by glenn-jocher** [comment]
Yes — that’s a sensible way to get one feature vector per GT from TAL-assigned positives, but `fg_mask` is not a strict “all anchors inside the original GT box” guarantee: as shown in [`select_candidates_in_gts()`](https://docs.ultralytics.com/reference/utils/tal/#ultralytics.utils.tal.TaskAlignedAssigner.select_candidates_in_gts), TAL uses anchor-center inclusion, then keeps only the assigned subset, and very small GT boxes may be expanded before selection, so your method is best interpreted as GT-assigned anchor pooling rather than exact within-box pooling.

**Comment by gigumay** [comment]
I see, thank you! I'm currently experimenting with aligning the extracted features with a prototype obtained by averaging the features of other models trained on the same classes but under different domains / data distributions. Essentially I measure the l2-distance/cosine similarity between the extracted features of a model and the prototype and then penalise training based on how different the representations are. It seems like this is making training more unstable and adding noise, though. Do you have an opinion on this approach? Am I maybe missing some architectural particularities of the YOLO model that make this impossible? In my mind, it should help models generalise better, but perhaps I'm wrong. Have you tried sth similar in the past?

**Comment by glenn-jocher** [comment]
That behavior makes sense to me: it’s not impossible, but YOLO’s FPN/TAL features are not trained to be a stable instance-embedding space, so pooling TAL-assigned anchors can be noisy and assignment-dependent; if you want prototype matching, I’d do it through a small projection head with normalized embeddings, a detached prototype, warmup, and a very small loss weight rather than penalizing raw detector features directly, and the [TAL reference](https://docs.ultralytics.com/reference/utils/tal/) plus [`DetectionPredictor.get_obj_feats()`](https://docs.ultralytics.com/reference/models/yolo/detect/predict/#ultralytics.models.yolo.detect.predict.DetectionPredictor.get_obj_feats) should help as implementation guides.

**Comment by gigumay** [comment]
What exactly do you mean by "detached prototype". To train the projection head I would have to backpropagate based on prototype similarity, so how can I detach it? Am I missing something?

**Comment by glenn-jocher** [comment]
By “detached prototype” I mean using the prototype as a fixed target on each step, e.g. `z = F.normalize(head(feats), dim=-1)`, `p = F.normalize(proto, dim=-1).detach()`, then minimize `1 - (z * p).sum(-1).mean()`: gradients still train the projection head through `z`, they just do not flow into `p`; if you want `p` to evolve, update it separately with an EMA or periodic recomputation rather than backpropagating through both sides at once, which usually makes training noisier.

**Comment by gigumay** [comment]
Thanks again! Which feature maps would you recommend I use? The ones coming out of the backbone (P3, P4, P5) or the ones coming out of the head (the ones that are passed to the `Detect` module)?

**Comment by glenn-jocher** [comment]
I’d start with the three feature maps fed into `Detect` rather than earlier backbone maps, since those neck/FPN features are the ones used by [the Detect head](https://docs.ultralytics.com/reference/nn/modules/head/) and [anchor generation](https://docs.ultralytics.com/reference/utils/tal/), so they match `fg_mask` and `target_gt_idx`
